sklearn/preprocessing/_encoders.py

```diff
@@ -0,0 +1,888 @@
+                "future release. You can use the 'categories' keyword instead."
+                " 'n_values=n' corresponds to 'categories=[range(n)]'.")
+            warnings.warn(msg, DeprecationWarning)
+
+            # we internally translate this to the correct categories
+            # and don't use legacy mode
+            X = check_array(X, dtype=np.int)
+
+            if isinstance(self._deprecated_n_values, numbers.Integral):
+                n_features = X.shape[1]
+                self.categories = [
+                    list(range(self._deprecated_n_values))
+                    for _ in range(n_features)]
+                n_values = np.empty(n_features, dtype=np.int)
+                n_values.fill(self._deprecated_n_values)
+            else:
+                try:
+                    n_values = np.asarray(self._deprecated_n_values, dtype=int)
+                    self.categories = [list(range(i))
+                                       for i in self._deprecated_n_values]
+                except (ValueError, TypeError):
+                    raise TypeError(
+                        "Wrong type for parameter `n_values`. Expected 'auto',"
+                        " int or array of ints, got %r".format(type(X)))
+
+            self._n_values_ = n_values
+            n_values = np.hstack([[0], n_values])
+            indices = np.cumsum(n_values)
+            self._feature_indices_ = indices
+
+            self._legacy_mode = False
+
+        else:  # n_values = 'auto'
+            if self.handle_unknown == 'ignore':
+                # no change in behaviour, no need to raise deprecation warning
+                self._legacy_mode = False
+            else:
+
+                # check if we have integer or categorical input
+                try:
+                    X = check_array(X, dtype=np.int)
+                except ValueError:
+                    self._legacy_mode = False
+                else:
+                    msg = (
+                        "The handling of integer data will change in the "
+                        "future. Currently, the categories are determined "
+                        "based on the range [0, max(values)], while in the "
+                        "future they will be determined based on the unique "
+                        "values.\nIf you want the future behaviour and "
+                        "silence this warning, you can specify "
+                        "\"categories='auto'\".\n"
+                        "In case you used a LabelEncoder before this "
+                        "OneHotEncoder to convert the categories to integers, "
+                        "then you can now use the OneHotEncoder directly."
+                    )
+                    warnings.warn(msg, DeprecationWarning)
+                    self._legacy_mode = True
+
+        # if user specified categorical_features -> always use legacy mode
+        if (not isinstance(self._deprecated_categorical_features,
```

Comment:
I would move this logic up. I expect nearly all users of ``OneHotEncoder`` so far have specified this.
That is if we indeed don't want to support this going forward. As I argued, I would rather keep this feature.

Reply:
I case we keep the deprecation, what do you mean with "move this logic up"? Inside this function? (and in that case, what would be the purpose?)